### PR TITLE
Run3-sim138 Changes which will allow the usage of new shower library (the cfi file in Geometry/HcalSimData will be modified once the new library appears)

### PR DESCRIPTION
**Pull request by bsunanda**
#### PR description:

Changes which will allow the usage of new shower library (the cfi file in Geometry/HcalSimData will be modified once the new library appears)

#### PR validation:

Use the runTheMatrix test workflows

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

Nothing special

### CODE REVIEW

**Comment by smuzaffar** [comment]
test parameters:
- pull_request = cms-sw/cmsdist#8292

**Comment by smuzaffar** [comment]
@bsunanda , feel free to update the cfi file to use the new shower library. https://github.com/cms-sw/cmssw/pull/40703#issuecomment-1418988477 show allow bot to pickup the new library when tests for this PR will run

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by smuzaffar** [comment]
please test

File: SimG4CMS/ShowerLibraryProducer/plugins/HcalForwardLibWriter.cc [hunk 1]
```diff
@@ -97,10 +99,15 @@ void HcalForwardLibWriter::analyze(const edm::Event& iEvent, const edm::EventSet
       ngood++;
       if (ngood > nshowers)
         continue;
+      unsigned int s = nphot;  //++
```
Comment (civanch) [inline SimG4CMS/ShowerLibraryProducer/plugins/HcalForwardLibWriter.cc:102]: @bsunanda , better not to use the name of variable "s", there may be clash with CLHEP.

**Comment by civanch** [comment]
@bsunanda , ideally in the plugin directory header and source files should be merged, also destructor should be implemented and not "default". I would agree if this PR is going forward and merged and in the next one this technical detail may be fixed.

**Comment by abdoulline** [comment]
Minor point:  probably it's worth also changing v6->v7  here
https://github.com/cms-sw/cmssw/blob/2a9ddc8a7cc0ea38ef7800661a87663f4be7abc2/Geometry/HcalSimData/python/HFParameters_cff.py#L35

**Comment by abdoulline** [comment]
I'm confused to see older versions (still in use for Run 1,2 eras) missing in the tests: 

 edm::FileInPath unable to find file SimG4CMS/Calo/data/HFShowerLibrary_oldpmt_noatt_eta4_16en_v3.root anywhere in the search path
 
edm::FileInPath unable to find file SimG4CMS/Calo/data/HFShowerLibrary_npmt_noatt_eta4_16en_v4.root anywhere in the search path

**Comment by smuzaffar** [comment]
my bad @abdoulline , I update the data package to only contain the new shower library. I will update https://github.com/cms-sw/cmsdist/pull/8292 to keep 
```
HFShowerLibrary_npmt_noatt_eta4_16en_v4.root
HFShowerLibrary_oldpmt_noatt_eta4_16en_v3.root
```
and only replace `HFShowerLibrary_run3_v6.root` with newer `HFShowerLibrary_run3_v7.root`

**Comment by smuzaffar** [comment]
please test

**Comment by civanch** [comment]
urgent

**Comment by civanch** [comment]
type urgent

**Comment by abdoulline** [comment]
@bsunanda 
Minor point:  I guess it's worth also changing v6->v7  here
https://github.com/cms-sw/cmssw/blob/2a9ddc8a7cc0ea38ef7800661a87663f4be7abc2/Geometry/HcalSimData/python/HFParameters_cff.py#L35

otherwise, in the absence of now **deleted** v6, customized Run2 jobs (to use Run3 SL) would fail...

**Comment by smuzaffar** [comment]
@bsunanda , look like https://github.com/cms-sw/cmssw/blob/master/SimG4CMS/Calo/test/test_catch2_hfshowerlibrary.cc#L183 also need to use the new v7 file

**Comment by perrotta** [comment]
urgent
- marking as "urgent" what was requested for 13_0_0_pre4 during the ORP meeting of Feb 7

**Comment by bsunanda** [comment]
Yes Run3 is supposed to use the new SL file while Runs 1 and 2 will use older files
________________________________

**Comment by civanch** [comment]
@bsunanda , @smuzaffar , new test of this PR should wait for the next IB, in which cmsdist will be updated? Are there any unit tests, which should be changed by hand in this PR?

**Comment by smuzaffar** [comment]
please test

**Comment by smuzaffar** [comment]
Please test

**Comment by smuzaffar** [comment]
@civanch , there are too many reco and dqm differences https://github.com/cms-sw/cmssw/pull/40703#issuecomment-1423667333 ,are these expected

**Comment by smuzaffar** [comment]
@civanch , looks like unit tests failure are not related to this PR but please check the dqm , reco comparison differences

**Comment by abdoulline** [comment]
There is something "behind the scene" = unrelated to this PR which spoils the DQM (comparison w/wo this PR) results.
For instance, in wf 10824.0_TTbar_13+2018  (a priori Run2 MC shouldn't be affected by this PR) there no any HCAL-related difference, while there are massive differences in Tracking (!) etc. 
https://cmsweb.cern.ch/dqm/dev/session/4JfrY8

**Comment by abdoulline** [comment]
But in Run3  wf 11634.0_TTbar_14TeV+2021, 
where HCAL differences appeared (as expected) and 
HF is supposed to be very similar (on average) to the reference,   
it doesn't seem to be the case... 
Mean energy (vs ieta) should be several GeV in HF,  while 
it's now close to 0...  
https://cmsweb.cern.ch/dqm/dev/plotfairy/overlay?session=jeHJFP;v=1675961702925253903;obj=archive%2F1%2FRelVal_wf11634_0_pr%2FCMSSW_13_0_X-PRcmssw_40703-55442%2FDQMIO%2FHcalRecHitsD%2FHcalRecHitTask%2Femean_vs_ieta_HF1;obj=archive%2F1%2FRelVal_wf11634_0_base%2FCMSSW_13_0_X-PRcmssw_40703-55442%2FDQMIO%2FHcalRecHitsD%2FHcalRecHitTask%2Femean_vs_ieta_HF1;reflabel=1;norm=False;w=1044;h=632

**Comment by abdoulline** [comment]
Hi @smuzaffar   
am I correct to assume that new (v7) HFShowerLibrary is available in the most recent IB 
CMSSW_13_0_X_2023-02-09-2300
to try to see what happens with this PR in some private test(s)?

**Comment by perrotta** [comment]
@bsunanda @abdoulline @smuzaffar @civanch  the evaluation of the effects of this PR looks stil far to converge: I think it should be descoped to 13_1_X. Do you envisage any serious counterindication with that?

**Comment by perrotta** [comment]
@bsunanda @abdoulline @smuzaffar @civanch  the evaluation of the effects of this PR looks stil far to converge: I think it should be descoped to 13_1_X. Do you envisage any serious counterindication with that?

**Comment by smuzaffar** [comment]
please test

lets rerun tests based on latest IB

**Comment by abdoulline** [comment]
I'm afraid Andrea is right, no need to run PR tests...
HF is **not OK** at all, as private single-pion test clearly shows:
https://cms-docs.web.cern.ch/hcal-sw-validation/calo_scan_single_pi/13_0_X/13_0_X_2023-02-09-2300_run3_PR40703_vs_13_0_0_pre3_run3_SinglePi/

**Comment by abdoulline** [comment]
Hi @kheyn 
There is clearly a problem with this PR (using your code + v7 SL). See my comments above...

**Comment by civanch** [comment]
Hi all, let us follow Andrea proposal. Whatever is the reason of bad results. 
Current v7 may be kept inside the data directory.

**Comment by kheyn** [comment]
Hi Salavat,

Could you forward me Andrea mail?

It would be good to have short chat.

Lev
________________________________

**Comment by abdoulline** [comment]
> Hi Salavat, Could you forward me Andrea mail? It would be good to have short chat. Lev

You've probably not realized you've actually replied to this PR chat...  - Andrea.Perrotta@cern.ch

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by abdoulline** [comment]
Thanks, Sunanda!
We've agreed with Vladimir and Sunanda to (try to) pursue PR submission, just reverting default (again to) v6.
It would help to debug the issue "in situ" and to introduce possibly very minimal corrections in the subsequent PR...

**Comment by civanch** [comment]
+1

by default the old v6 version of SL is used, so there is full regression and this PR may be merged allowing further investigation of v7.

**Comment by rappoccio** [comment]
hold

**Comment by rappoccio** [comment]
This is on hold because it is fully signed but it seems it is not resolved.

**Comment by civanch** [comment]
@rappoccio , in our discussion we agreed to merge this PR and resolve problem in the following PR. Technically it is easier, especially because PR was moved from 13_0 to 13_1. This PR includes reasonable update of the current code and do not provide any problem to any WF.

**Comment by perrotta** [comment]
@rappoccio commit https://github.com/cms-sw/cmssw/pull/40703/commits/f09815aaf961878ea4c2b4de8495a5e146201b09 reverts the version of the HF shower library to previous version 2: therefore as of now this PR does not affect any workflow, but it allows studies on that version of the shower library to be performed in the development release

**Comment by perrotta** [comment]
unhold

**Comment by perrotta** [comment]
+1

**Pull request merged**